docs: describe workspace tags (#13352)
coder/coder: docs/templates/workspace-tags.md

# Workspace Tags

> Template administrators can use static template tags to restrict workspace
> provisioning to specific provisioner groups. However, this method has limited
> flexibility as it prevents workspace users from creating workspaces on nodes of
> their choice.

Reviewer: ```suggestion
Template administrators can use static template tags to restrict workspace
provisioning to specific provisioner groups. However, this method has limited
flexibility and lacks the capability to allow users to choose the nodes on which
they create workspaces.
```
 It still looks fine, but I would replace `prevent` with something like `lacks.`

By using `coder_workspace_tags` and `coder_parameter`s, template administrators
can enable dynamic tag selection and modify static template tags.

## Dynamic tag selection

Here is a sample `coder_workspace_tags` data resource with a few workspace tags
specified:

```hcl
data "coder_workspace_tags" "custom_workspace_tags" {
  tags = {
    "zone"        = "developers"
    "os"          = data.coder_parameter.os_selector.value
    "project_id"  = "PROJECT_${data.coder_parameter.project_name.value}"
    "cache"       = data.coder_parameter.feature_cache_enabled.value == "true" ? "with-cache" : "no-cache"
  }
}
```

**Legend**

- `zone` - static tag value set to `developers`
- `os` - supported by the string-type `coder_parameter` to select OS
  runtime,`os_selector`
- `project_id` - a formatted string supported by the string-type
  `coder_parameter`, `project_name`
- `cache` - an HCL condition involving boolean-type `coder_parameter`,
  `feature_cache_enabled`

Review the
[full template example](https://github.com/coder/coder/tree/main/examples/workspace-tags)
using `coder_workspace_tags` and `coder_parameter`s.

## Constraints

### Tagged provisioners

> With incorrectly selected workspace tags, it is possible to choose a tag
> configuration that is not observed by any provisioner, causing the provisioner
> job to get stuck in the queue indefinitely.

Reviewer 2: ```suggestion
It is possible to choose tag combinations that no provisioner can handle.
This will cause the provisioner job to get stuck in the queue until
a provisioner is added that can handle its combination of tags.
```
